// node-term-ui/TermUI.js
(function() {
  var Button, EventEmitter, T, TermUI, Widget, b1, b2, b3, b4, b5, b6, b7, b8, b9, log, stdio, tty, util, _;
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; }, __hasProp = Object.prototype.hasOwnProperty, __extends = function(child, parent) {
    for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; }
    function ctor() { this.constructor = child; }
    ctor.prototype = parent.prototype;
    child.prototype = new ctor;
    child.__super__ = parent.prototype;
    return child;
  };
  util = require('util');
  tty = require('tty');
  stdio = process.binding('stdio');
  EventEmitter = require('events').EventEmitter;
  _ = require('underscore');
  _.mixin(require('underscore.string'));
  log = console.log;
  TermUI = (function() {
    __extends(TermUI, EventEmitter);
    function TermUI() {
      this.handleData = __bind(this.handleData, this);
      this.handleKeypress = __bind(this.handleKeypress, this);
      this.handleSizeChange = __bind(this.handleSizeChange, this);      if (tty.isatty(process.stdin)) {
        tty.setRawMode(true);
        process.stdin.resume();
        process.stdin.on('keypress', this.handleKeypress);
        process.stdin.on('data', this.handleData);
        if (process.listeners('SIGWINCH').length === 0) {
          process.on('SIGWINCH', this.handleSizeChange);
        }
        this.handleSizeChange();
        this.enableMouse();
        this.isTerm = true;
      } else {
        this.isTerm = false;
      }
    }
    TermUI.prototype.C = {
      k: 0,
      r: 1,
      g: 2,
      y: 3,
      b: 4,
      m: 5,
      c: 6,
      w: 7,
      x: 9
    };
    TermUI.prototype.S = {
      normal: 0,
      bold: 1,
      underline: 4,
      blink: 5,
      inverse: 8
    };
    TermUI.prototype.SYM = {
      star: '\u2605',
      check: '\u2714',
      x: '\u2718',
      triUp: '\u25b2',
      triDown: '\u25bc',
      triLeft: '\u25c0',
      triRight: '\u25b6',
      fn: '\u0192',
      arrowUp: '\u2191',
      arrowDown: '\u2193',
      arrowLeft: '\u2190',
      arrowRight: '\u2192'
    };
    TermUI.prototype.handleSizeChange = function() {
      var winsize;
      winsize = tty.getWindowSize(this.stdout);
      this.width = winsize[1];
      this.height = winsize[0];
      return this.emit('resize', {
        w: this.width,
        h: this.height
      });
    };
    TermUI.prototype.out = function(buf) {
      if (this.isTerm) {
        process.stdout.write(buf);
      }
      return this;
    };
    TermUI.prototype.clear = function() {
      this.out('\x1b[2J');
      this.home;
      return this;
    };
    TermUI.prototype.pos = function(x, y) {
      x = x < 0 ? this.width - x : x;
      y = y < 0 ? this.height - y : y;
      x = Math.max(Math.min(x, this.width), 1);
      y = Math.max(Math.min(y, this.height), 1);
      this.out("\x1b[" + y + ";" + x + "H");
      return this;
    };
    TermUI.prototype.home = function() {
      this.pos(1, 1);
      return this;
    };
    TermUI.prototype.end = function() {
      this.pos(1, -1);
      return this;
    };
    TermUI.prototype.fg = function(c) {
      this.out("\x1b[3" + c + "m");
      return this;
    };
    TermUI.prototype.bg = function(c) {
      this.out("\x1b[4" + c + "m");
      return this;
    };
    TermUI.prototype.hifg = function(c) {
      this.out("\x1b[38;5;" + c + "m");
      return this;
    };
    TermUI.prototype.hibg = function(c) {
      this.out("\x1b[48;5;" + c + "m");
      return this;
    };
    TermUI.prototype.enableMouse = function() {
      this.out('\x1b[?1000h');
      this.out('\x1b[?1002h');
      return this;
    };
    TermUI.prototype.disableMouse = function() {
      this.out('\x1b[?1000l');
      this.out('\x1b[?1002l');
      return this;
    };
    TermUI.prototype.eraseLine = function() {
      this.out('\x1b[2K');
      return this;
    };
    TermUI.prototype.handleKeypress = function(c, key) {
      if (key && key.ctrl && key.name === 'c') {
        console.log("ctrlc");
        return this.quit();
      } else {
        return this.emit('keypress', c, key);
      }
    };
    TermUI.prototype.handleData = function(d) {
      var buttons, event, eventData;
      eventData = {};
      buttons = ['left', 'middle', 'right'];
      if (d[0] === 0x1b && d[1] === 0x5b && d[2] === 0x4d) {
        switch (d[3] & 0x60) {
          case 0x20:
            if ((d[3] & 0x3) < 0x3) {
              event = 'mousedown';
              eventData.button = buttons[d[3] & 0x3];
            } else {
              event = 'mouseup';
            }
            break;
          case 0x40:
            event = 'drag';
            if ((d[3] & 0x3) < 0x3) {
              eventData.button = buttons[d[3] & 0x3];
            }
            break;
          case 0x60:
            event = 'wheel';
            if (d[3] & 0x1) {
              eventData.direction = 'down';
            } else {
              eventData.direction = 'up';
            }
        }
        eventData.shift = (d[3] & 0x4) > 0;
        eventData.x = d[4] - 32;
        eventData.y = d[5] - 32;
        this.emit(event, eventData);
        return this.emit('any', event, eventData);
      }
    };
    TermUI.prototype.quit = function() {
      this.fg(this.C.x).bg(this.C.x);
      this.disableMouse();
      tty.setRawMode(false);
      return process.exit();
    };
    return TermUI;
  })();
  T = exports.TermUI = new TermUI;
  T.enableMouse();
  Widget = (function() {
    __extends(Widget, EventEmitter);
    function Widget(options) {
      var _ref, _ref2, _ref3, _ref4;
      this.options = options != null ? options : {};
      this.bounds = {
        x: ((_ref = this.options.bounds) != null ? _ref.x : void 0) || 0,
        y: ((_ref2 = this.options.bounds) != null ? _ref2.y : void 0) || 0,
        w: ((_ref3 = this.options.bounds) != null ? _ref3.w : void 0) || 0,
        h: ((_ref4 = this.options.bounds) != null ? _ref4.h : void 0) || 0
      };
      Widget.instances.unshift(this);
    }
    Widget.prototype.draw = function() {};
    Widget.prototype.hitTest = function(x, y) {
      return ((this.bounds.x <= x && x <= (this.bounds.x + this.bounds.w - 1))) && ((this.bounds.y <= y && y <= (this.bounds.y + this.bounds.h - 1)));
    };
    return Widget;
  })();
  Widget.instances = [];
  T.on('any', function(event, eventData) {
    var widget, _i, _len, _ref, _results;
    _ref = Widget.instances;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      widget = _ref[_i];
      _results.push(widget.hitTest(eventData.x, eventData.y) ? (eventData.target = widget, widget.emit(event, eventData)) : void 0);
    }
    return _results;
  });
  Button = (function() {
    __extends(Button, Widget);
    function Button(opts) {
      var _ref, _ref2, _ref3, _ref4;
      Button.__super__.constructor.call(this, opts);
      this.fg = (_ref = this.options.fg) != null ? _ref : T.C.w;
      this.bg = (_ref2 = this.options.fg) != null ? _ref2 : T.C.b;
      this.label = (_ref3 = this.options.label) != null ? _ref3 : '';
      this.labelAnchor = (_ref4 = this.options.labelAnchor) != null ? _ref4 : 5;
    }
    Button.prototype.draw = function() {
      var align, emptyStr, labelRow, labelStr, y, _ref, _ref2;
      T.fg(this.fg).bg(this.bg).pos(this.bounds.x, this.bounds.y);
      align = ['lpad', 'rpad', 'center'][this.labelAnchor % 3];
      labelStr = _[align](this.label, this.bounds.w, ' ');
      if (this.bounds.h > 1) {
        emptyStr = _.pad(' ', this.bounds.w, ' ');
      }
      switch (((this.labelAnchor - 1) / 3) | 0) {
        case 0:
          labelRow = this.bounds.y + this.bounds.h - 1;
          break;
        case 1:
          labelRow = this.bounds.y + (this.bounds.h / 2) | 0;
          break;
        case 2:
          labelRow = this.bounds.y;
      }
      for (y = _ref = this.bounds.y, _ref2 = this.bounds.y + this.bounds.h - 1; _ref <= _ref2 ? y <= _ref2 : y >= _ref2; _ref <= _ref2 ? y++ : y--) {
        T.pos(this.bounds.x, y);
        if (y === labelRow) {
          T.out(labelStr);
        } else {
          T.out(emptyStr);
        }
      }
      return T.fg(T.C.x).bg(T.C.x).end();
    };
    return Button;
  })();
  exports.Widget = Widget;
  exports.Button = Button;
  if (process.argv[1] === __filename) {
    T.clear();
    b7 = new Button({
      bounds: {
        x: 1,
        y: 1,
        w: 20,
        h: 3
      },
      label: 'hello 7',
      labelAnchor: 7
    });
    b7.draw();
    b8 = new Button({
      bounds: {
        x: 22,
        y: 1,
        w: 20,
        h: 3
      },
      label: 'hello 8',
      labelAnchor: 8
    });
    b8.draw();
    b9 = new Button({
      bounds: {
        x: 43,
        y: 1,
        w: 20,
        h: 3
      },
      label: 'hello 9',
      labelAnchor: 9
    });
    b9.draw();
    b4 = new Button({
      bounds: {
        x: 1,
        y: 5,
        w: 20,
        h: 3
      },
      label: 'hello 4',
      labelAnchor: 4
    });
    b4.draw();
    b5 = new Button({
      bounds: {
        x: 22,
        y: 5,
        w: 20,
        h: 3
      },
      label: 'hello 5',
      labelAnchor: 5
    });
    b5.draw();
    b5.on('mousedown', function() {
      b5.bg = T.C.y;
      return b5.draw();
    });
    b5.on('mouseup', function() {
      b5.bg = T.C.b;
      return b5.draw();
    });
    b6 = new Button({
      bounds: {
        x: 43,
        y: 5,
        w: 20,
        h: 3
      },
      label: 'hello 6',
      labelAnchor: 6
    });
    b6.draw();
    b1 = new Button({
      bounds: {
        x: 1,
        y: 9,
        w: 20,
        h: 3
      },
      label: 'hello 1',
      labelAnchor: 1
    });
    b1.draw();
    b2 = new Button({
      bounds: {
        x: 22,
        y: 9,
        w: 20,
        h: 3
      },
      label: 'hello 2',
      labelAnchor: 2
    });
    b2.draw();
    b3 = new Button({
      bounds: {
        x: 43,
        y: 9,
        w: 20,
        h: 3
      },
      label: 'hello 3',
      labelAnchor: 3
    });
    b3.draw();
  }
}).call(this);
